File: FinalProject/1_IntegerPartition/Baitoan3/PkSelfCjg.cpp
#include <bits/stdc++.h>
using namespace std;

bool isSelfConjugate(const vector<int>& parts) {
    int m = (int)parts.size();
    vector<int> conjugate;
    int maxPart = 0;
    for (int x : parts) maxPart = max(maxPart, x);

    for (int i = 1; i <= maxPart; i++) {
        int cnt = 0;
        for (int val : parts) {
            if (val >= i) cnt++;
            else break;
        }
        conjugate.push_back(cnt);
    }

    if (conjugate.size() != parts.size()) return false;
    for (int i = 0; i < (int)parts.size(); i++) {
        if (parts[i] != conjugate[i]) return false;
    }
    return true;
}

// (i) đệ quy
map<pair<int,int>, int> memo_recursive;

int p_selfcjg_recursive(int n, int k) {
    if (n == 0 && k == 0) return 1;
    if (n < 0 || k <= 0) return 0;
    if (k > n) return 0;

    pair<int,int> key = {n, k};
    if (memo_recursive.count(key)) return memo_recursive[key];

    int count = 0;
    vector<int> current;
    // Sinh phân hoạch có k phần tổng n theo maxVal
    function<void(int,int,int)> generate = [&](int remain, int parts, int maxVal) {
        if (parts == 0 && remain == 0) {
            vector<int> temp = current;
            sort(temp.rbegin(), temp.rend());
            if ((int)temp.size() == k && isSelfConjugate(temp)) {
                count++;
            }
            return;
        }
        if (parts <= 0 || remain <= 0) return;
        for (int val = min(maxVal, remain); val >= 1; --val) {
            current.push_back(val);
            generate(remain - val, parts - 1, val);
            current.pop_back();
        }
    };

    generate(n, k, n);
    return memo_recursive[key] = count;
}

// (ii) quy hoạch động
int p_selfcjg_dp(int n, int k) {
    if (n == 0 && k == 0) return 1;
    if (n < 0 || k <= 0) return 0;
    if (k > n) return 0;

    // dp[i][j] = số phân hoạch tự liên hợp của i có j phần
    vector<vector<int>> dp(n+1, vector<int>(k+1, 0));
    dp[0][0] = 1;

    for (int sum = 1; sum <= n; ++sum) {
        for (int parts = 1; parts <= min(sum, k); ++parts) {
            int count = 0;
            vector<int> current;

            function<void(int,int,int)> generate = [&](int remain, int currParts, int maxVal) {
                if (currParts == 0 && remain == 0) {
                    vector<int> temp = current;
                    sort(temp.rbegin(), temp.rend());
                    if ((int)temp.size() == parts && isSelfConjugate(temp)) {
                        count++;
                    }
                    return;
                }
                if (currParts <= 0 || remain <= 0) return;
                for (int val = min(maxVal, remain); val >= 1; --val) {
                    current.push_back(val);
                    generate(remain - val, currParts - 1, val);
                    current.pop_back();
                }
            };

            generate(sum, parts, sum);
            dp[sum][parts] = count;
        }
    }

    return dp[n][k];
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    int n, k;
    cin >> n >> k;

    memo_recursive.clear();
    int resRec = p_selfcjg_recursive(n, k);
    cout << "So phan hoach tu lien hop cua " << n << " co " << k << " phan (De quy): " << resRec;

    int resDP = p_selfcjg_dp(n, k);
    cout << "\nSo phan hoach tu lien hop cua " << n << " co " << k << " phan (DP): " << resDP;

    return 0;
}

File: FinalProject/1_IntegerPartition/Baitoan3/PkSelfCjg.py
from typing import List

def is_self_conjugate(parts: List[int]) -> bool:
    m = len(parts)
    conjugate = []
    max_part = max(parts) if parts else 0

    for i in range(1, max_part + 1):
        cnt = 0
        for val in parts:
            if val >= i:
                cnt += 1
            else:
                break
        conjugate.append(cnt)

    if len(conjugate) != len(parts):
        return False
    for i in range(len(parts)):
        if parts[i] != conjugate[i]:
            return False
    return True


# (i) Đệ quy
memo_recursive = {}

def p_selfcjg_recursive(n, k):
    if n == 0 and k == 0:
        return 1
    if n < 0 or k <= 0 or k > n:
        return 0

    key = (n, k)
    if key in memo_recursive:
        return memo_recursive[key]

    count = 0
    current = []

    def generate(remain, parts, max_val):
        nonlocal count
        if parts == 0 and remain == 0:
            temp = sorted(current, reverse=True)
            if len(temp) == k and is_self_conjugate(temp):
                count += 1
            return
        if parts <= 0 or remain <= 0:
            return
        for val in range(min(max_val, remain), 0, -1):
            current.append(val)
            generate(remain - val, parts - 1, val)
            current.pop()

    generate(n, k, n)
    memo_recursive[key] = count
    return count


# (ii) Quy hoạch động
def p_selfcjg_dp(n, k):
    if n == 0 and k == 0:
        return 1
    if n < 0 or k <= 0 or k > n:
        return 0

    dp = [[0] * (k + 1) for _ in range(n + 1)]
    dp[0][0] = 1

    for total in range(1, n + 1):
        for parts in range(1, min(k, total) + 1):
            count = 0
            current = []

            def generate(remain, curr_parts, max_val):
                nonlocal count
                if curr_parts == 0 and remain == 0:
                    temp = sorted(current, reverse=True)
                    if len(temp) == parts and is_self_conjugate(temp):
                        count += 1
                    return
                if curr_parts <= 0 or remain <= 0:
                    return
                for val in range(min(max_val, remain), 0, -1):
                    current.append(val)
                    generate(remain - val, curr_parts - 1, val)
                    current.pop()

            generate(total, parts, total)
            dp[total][parts] = count

    return dp[n][k]

n, k = map(int, input().split())

memo_recursive.clear()
res_rec = p_selfcjg_recursive(n, k)
print(f"Số phân hoạch tự liên hợp của {n} có {k} phần (Đệ quy): {res_rec}")

res_dp = p_selfcjg_dp(n, k)
print(f"Số phân hoạch tự liên hợp của {n} có {k} phần (DP): {res_dp}")